spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/RetryLoadBalancerInterceptor.java

```diff
@@ -74,19 +88,35 @@ public ClientHttpResponse intercept(final HttpRequest request, final byte[] body
 				LoadBalancedRetryContext lbContext = (LoadBalancedRetryContext) context;
 				serviceInstance = lbContext.getServiceInstance();
 			}
+			Set<LoadBalancerLifecycle> supportedLifecycleProcessors = supportedLifecycleProcessors(
+					serviceName);
+			String hint = getHint(serviceName);
+			DefaultRequest<HttpRequestContext> lbRequest = new DefaultRequest<>(
+					new HttpRequestContext(request, hint));
+			supportedLifecycleProcessors
+					.forEach(lifecycle -> lifecycle.onStart(lbRequest));
 			if (serviceInstance == null) {
-				serviceInstance = this.loadBalancer.choose(serviceName);
+				serviceInstance = this.loadBalancer.choose(serviceName, lbRequest);
 			}
+			Response<ServiceInstance> lbResponse = new DefaultResponse(serviceInstance);
 			ClientHttpResponse response = RetryLoadBalancerInterceptor.this.loadBalancer
 					.execute(serviceName, serviceInstance,
 							this.requestFactory.createRequest(request, body, execution));
 			int statusCode = response.getRawStatusCode();
 			if (retryPolicy != null && retryPolicy.retryableStatusCode(statusCode)) {
 				byte[] bodyCopy = StreamUtils.copyToByteArray(response.getBody());
 				response.close();
-				throw new ClientHttpResponseStatusCodeException(serviceName, response,
-						bodyCopy);
+				ClientHttpResponseStatusCodeException clientHttpResponseStatusCodeException = new ClientHttpResponseStatusCodeException(
+						serviceName, response, bodyCopy);
+				supportedLifecycleProcessors.forEach(lifecycle -> lifecycle.onComplete(
```

Comment:
I'm not entirely sure whether we should call `onComplete` here as lifecycle processors' `onComplete` from `BlockingLoadBalancerClient` will also be called during the `execute` method call in line 103. However, in `BlockingLoadBalancerClient` we pass `DefaultRequestContext` and `Object` for support check and here `HttpRequestContext` and `ClientHttpResponse`, so we can have more granular lifecycle processors. On the other hand, the ones that match with `DefaultRequestContext` and `Object` will be executed twice, which is not optimal.
@spencergibb wdyt?

Reply:
shouldn't we call `loadBalancer.execute()`? Then that will take care of the lifecycle? We should test that lifecycle methods aren't called twice. Not required to merge, but should happen shortly after.

Reply:
Regarding using `execute()`, I have added [this comment](https://github.com/spring-cloud/spring-cloud-commons/pull/783#discussion_r458025632). However, actually, as a result of [this call](https://github.com/spring-cloud/spring-cloud-commons/blob/c8a57be0f1d82f9f381636ac530ad69fc0985a9e/spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/RetryLoadBalancerInterceptor.java#L80-L82), the Lifecycle `onComplete()` calls for `FAILED` and `SUCCESS` are being executed anyway. I thought that adding them here would allow for a better granularity since here we can filter the lifecycle processors by `ClientHttpResponse.class` as opposed to `Object.class` in `BlockingLoadBalancerClient`. However, as they would duplicate, I will remove it from here and add the test you've suggested. I will also add calling `onComplete(DISCARDED[...])` for when the retrieved service instance is `null`. Like this, all calls should be handled. Let me know what you think about this.